Skip generated column verification if value is not provided

BaurzhanSakhariev opened the pull request:
No need to check generated value with itself.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dml/ObjectIndexer.java:
```diff
@@ -96,11 +96,13 @@ public void indexValue(@NotNull Map<String, Object> value, IndexDocumentBuilder
             Child child = entry.getValue();
             Object innerValue = null;
             if (value.containsKey(innerName) == false) {
+                // Generated value is not provided, nothing to check, use generated value.
                 innerValue = docBuilder.getSyntheticValue(child.ident());
```
Actually, I meant to target only generated value verification. 
This is column constraint, so we are not omitting CHECK constraint. 

Looks like NOT NULL constraint got skipped, will double check.

BaurzhanSakhariev replied:
```
execute("create table t (id int, obj object as (a object as (b int not null)))");
execute("insert into t (id, obj) (select 1, {\"a\" = {}} from sys.cluster)");
assertThat(response.rowCount()).isEqualTo(0L);
```

works both on master and on this PR. This is because NOT NULL is triggered as a table constraint because it's not in target columns. Looks like in general it's not possible to have sub-column column constraint NOT NULL but I will push a change to target only generated column, we anyway don't want to skip NOT NULL check

BaurzhanSakhariev replied:
Pushed a change to be more future-proof. 

I can't think of example where first commit state would be failing but let's keep it safe and skip only generated not provided.

BaurzhanSakhariev commented:
I confirmed that running reproduction script from https://github.com/crate/support/issues/423 on the build with this change makes final target table contain records.

If I run repro script against master build, I don't see OOM but target table is empty.

BaurzhanSakhariev commented:
Hi @mfussenegger can you please take a look? 5.9.0 is done, this should be fine to go

mfussenegger commented on server/src/main/java/io/crate/execution/dml/IndexDocumentBuilder.java:
```diff
@@ -89,10 +89,12 @@ public Object getSyntheticValue(ColumnIdent columnIdent) {
+    public void checkColumnConstraint(ColumnIdent columnIdent, Object value, boolean noValue) {
         Indexer.ColumnConstraint constraint = constraints.get(columnIdent);
-        if (constraint != null) {
+        if (constraint != null && (noValue && constraint instanceof Indexer.CheckGeneratedValue) == false) {
```
Afaik the constraint could be a `MultiCheck`.

I'm generally not a fan of this approach. I'd rather fix the root cause.

BaurzhanSakhariev replied:
This is basically micro-optimization and just happens to be a bug fix.

Making `GeoShapeType.compareTo ` correct can fix it but then for inserting big generated Polygons with hundreds of points we would compare generated polygon with itself(!) per row. 

I agree that the root cause must be fixed but still think this is useful in general. 
Generated columns without provided value is a very common use case.

mfussenegger replied:
> And we definitely should fix the root cause first, otherwise we're just masking bugs

To elaborate on this: if the user provides the values which match the generated value they'd still fail if we don't fix the root cause

BaurzhanSakhariev commented:
Fix of the root cause is done in https://github.com/crate/crate/pull/16727

I will rework this as an optimization later and will address

> Then we should find an approach that doesn't break encapsulation.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -334,9 +334,9 @@ public void verify(Object[] values) {
         @Override
-        public void verify(Object providedValue) {
+        public void verify(Object providedValue, boolean noValue) {
             for (var check : checks) {
-                check.verify(providedValue);
+                check.verify(providedValue, noValue);
```
MultiCheck - skips only generated but keeps other checks

BaurzhanSakhariev commented:
Hi @jeeminso, can you please take a look?

Basically, this change is about not checking generated value with itself.

If value  is not provided, generated value check is redundant.

I addressed comments above and got rid of instance of checks - instead just passed noValue flag further down so that GeneratedValueRef can skip it (also when part of a MultiCheck) and otherwise all other checks behave as usual.

jeeminso commented on server/src/main/java/io/crate/execution/dml/ObjectIndexer.java:
```diff
@@ -110,7 +112,7 @@ public void indexValue(@NotNull Map<String, Object> value, IndexDocumentBuilder
                 }
             }
             var innerValue = value.get(innerName);
-            docBuilder.checkColumnConstraint(child.ident(), innerValue);
+            docBuilder.checkColumnConstraint(child.ident(), innerValue, noValue);
```
Can't we clean up further? It looks like the `if` block right above is the only place that will ever call `checkColumnConstraint()` or `verify()` methods with `noValue = true`. Can we simplify the whole PR by wrapping these two lines in an `else` block?

BaurzhanSakhariev replied:
That was one of my first drafts in https://github.com/crate/crate/commit/8b33266f7f5ebd0049ea4378763fdb96419af28b
but we can't skip the whole check only if value is not provided - we still need to apply CHECK/MultiChecK/NotNull and their are hidden in `checkColumnConstraint`

jeeminso reviewed (approved):
Thanks, LGTM. Just one minor suggestion to double check.

BaurzhanSakhariev commented:
@Mergifyio refresh
